### PR TITLE
*: add system table `mysql.stats_meta`

### PR DESCRIPTION
**Pull request by hanfei1991**
We add a new system table  `mysql.stats_meta` to record table stats version.
In futher, this table will record the cardinality and distinct value of a table.
We load stats info in a loop. Everytime we just load the stats whose version is greator than last stats version.
Everytime we analyze a table, we update the version of that table stats.
@shenli @coocood @zimulala @tiancaiamao @lamxTyler PTAL

### CODE REVIEW

File: bootstrap.go [hunk 1]
```diff
@@ -115,6 +115,15 @@ const (
   		PRIMARY KEY (help_topic_id),
   		UNIQUE KEY name (name)
 		) ENGINE=InnoDB DEFAULT CHARSET=utf8 STATS_PERSISTENT=0 COMMENT='help topics';`
+
+	// CreateStatsMetaTable store's the meta of table statistics.
+	CreateStatsMetaTable = `CREATE TABLE if not exists mysql.stats_meta (
+		version bigint(64) unsigned NOT NULL,
+		table_id bigint(64) NOT NULL,
+		cardinality int(64) unsigned NOT NULL DEFAULT 0,
+		dist int(64) unsigned NOT NULL DEFAULT 0,
```
Comment (coocood) [inline bootstrap.go]: cardinality is distince count

File: plan/statscache/statscache.go [hunk 2]
```diff
@@ -14,89 +14,109 @@
...
-		tbl:      statsTbl,
-		loadTime: oracle.GetPhysical(time.Now()),
+func SetStatisticsTableCache(id int64, statsTbl *statistics.Table, version uint64) {
+	stats, ok := statsTblCache.cache[id]
+	if !ok {
+		si := &statsInfo{
+			tbl:     statsTbl,
+			version: version,
+		}
+		statsTblCache.cache[id] = si
```
Comment (lamxTyler) [inline plan/statscache/statscache.go]: Is this concurrency safe?

File: plan/statscache/statscache_test.go [hunk 3]
```diff
@@ -33,44 +35,41 @@ var _ = Suite(&testStatsCacheSuite{})
...
+	c.Assert(statsTbl.Pseudo, IsTrue)
+	testKit.MustExec("analyze table t")
+	statsTbl = statscache.GetStatisticsTableCache(tableInfo)
+	c.Assert(statsTbl.Pseudo, IsFalse)
+	testKit.MustExec("create index idx_t on t(c1)")
+	is = do.InfoSchema()
+	tbl, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tbl, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
```
Comment (lamxTyler) [inline plan/statscache/statscache_test.go]: Those two lines is the same as the 56-57 lines.

File: plan/statscache/statscache.go [hunk 4]
```diff
@@ -14,89 +14,109 @@
...
+	return nil
+}
+
+type statsCache struct {
+	cache map[int64]*statsInfo
+}
+
+var statsTblCache = statsCache{cache: map[int64]*statsInfo{}}
+
+// GetStatisticsTableCache retrieves the statistics table from cache, and will refill cache if necessary.
```
Comment (lamxTyler) [inline plan/statscache/statscache.go]: Comment should be updated, it doesn't refill cache now.

File: plan/statscache/statscache.go [hunk 5]
```diff
@@ -14,89 +14,109 @@
...
+		tableInfo := table.Meta()
+		tpb, err := m.GetTableStats(tableID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		var tbl *statistics.Table
+		if tpb != nil {
+			tbl, err = statistics.TableFromPB(tableInfo, tpb)
+			// Error is not nil may mean that there are some ddl changes on this table, so the origin
+			// statistics can not be used any more, we give it a pseudo one and save in cache.
```
Comment (lamxTyler) [inline plan/statscache/statscache.go]: Here the pseudo one is not easily see, keep it nil is not the same as give it a pseudo one.

Comment (hanfei1991) [inline plan/statscache/statscache.go]: If there is a nil stats, we always return a pseudo one.

File: domain/domain.go [hunk 6]
```diff
@@ -398,6 +407,48 @@ func (do *Domain) PrivilegeHandle() *privileges.Handle {
...
+func (do *Domain) LoadTableStatsLoop(ctx context.Context) error {
+	do.statsHandle = statscache.NewHandle(ctx)
+	err := do.loadTableStats()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	lease := do.DDL().GetLease()
+	if lease > 0 {
+		go func(do *Domain) {
+			ticker := time.NewTicker(lease)
```
Comment (tiancaiamao) [inline domain/domain.go]: Why the reload interval equal to a lease?

File: executor/analyze.go [hunk 7]
```diff
@@ -125,6 +128,16 @@ func (e *AnalyzeExec) buildStatisticsAndSaveToKV(count int64, columnSamples [][]
 	if err != nil {
 		return errors.Trace(err)
 	}
+	delSQL := fmt.Sprintf("delete from mysql.stats_meta where table_id = %d", e.tblInfo.ID)
+	_, _, err = e.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(e.ctx, delSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	insertSQL := fmt.Sprintf("insert into mysql.stats_meta (version, table_id) values(%d, %d)", version, e.tblInfo.ID)
```
Comment (tiancaiamao) [inline executor/analyze.go]: add a whitespace between `values(%d, %d)` ?

File: plan/statscache/statscache.go [hunk 8]
```diff
@@ -14,89 +14,109 @@
...
-	statsTblCache.m.RLock()
-	si, ok := statsTblCache.cache[tblInfo.ID]
-	statsTblCache.m.RUnlock()
-	// Here we check the TableInfo because there may be some ddl changes in the duration period.
-	// Also, we rely on the fact that TableInfo will not be same if and only if there are ddl changes.
-	if ok && tblInfo == si.tbl.Info && !tableCacheExpired(si) {
-		return si.tbl
+// Update reads stats meta from store and updates the stats map.
+func (h *Handle) Update(m *meta.Meta, is infoschema.InfoSchema) error {
+	sql := fmt.Sprintf("SELECT version, table_id from mysql.stats_meta where version > %d order by version", h.lastVersion)
```
Comment (tiancaiamao) [inline plan/statscache/statscache.go]: I'm not quite understand about the version, do you mean statistics information is out of date when infoschema changed ?

File: plan/statscache/statscache.go (hunk 2 above)
Comment (tiancaiamao) [inline plan/statscache/statscache.go]: `statsTblCache.m.Lock()` already called.  @lamxTyler

File: session.go [hunk 9]
```diff
@@ -114,6 +114,9 @@ type session struct {
 	parser    *parser.Parser
 
 	sessionVars *variable.SessionVars
+
+	// sysSessionPool stores sessions for execute restricted sql for every domain.
+	sysSessionPool *sync.Pool
```
Comment (tiancaiamao) [inline session.go]: Why `sysSessionPool` in so many place? It's hard to maintain.

File: domain/domain.go (hunk 6 above)
Comment (hanfei1991) [inline domain/domain.go]: When a user execute analyze, we must notify others as soon as possible. I think a lease is ok.

File: bootstrap.go [hunk 10]
```diff
@@ -298,6 +307,8 @@ func doDDLWorks(s Session) {
 	mustExecute(s, CreateTiDBTable)
 	// Create help table.
 	mustExecute(s, CreateHelpTopic)
+	// Create stats_meta table.
+	mustExecute(s, CreateStatsMetaTable)
```
Comment (hanfei1991) [inline bootstrap.go]: Yeah, I will add the upgrade logic soon.

File: plan/statscache/statscache.go (hunk 8 above)
Comment (hanfei1991) [inline plan/statscache/statscache.go]: stats info version will increase when the table is analyzed. The version is `analyze command`'s startTS.

File: session.go (hunk 9 above)
Comment (hanfei1991) [inline session.go]: They all be fetched from same domain. Otherwise which place is ok?

Comment (tiancaiamao) [inline session.go]: it's easy to define a function such as 

```
func (s *session) sysSessionPool() *sync.Pool {
    return sessionctx.GetDomain(s).SysSessionPool()
}
```

Comment (tiancaiamao) [inline session.go]: Rest LGTM
@coocood @lamxTyler

**Comment by lamxTyler** [comment]
LGTM

**Comment by hanfei1991** [comment]
@tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
@coocood

File: bootstrap.go [hunk 11]
```diff
@@ -115,6 +115,15 @@ const (
   		PRIMARY KEY (help_topic_id),
   		UNIQUE KEY name (name)
 		) ENGINE=InnoDB DEFAULT CHARSET=utf8 STATS_PERSISTENT=0 COMMENT='help topics';`
+
+	// CreateStatsMetaTable store's the meta of table statistics.
+	CreateStatsMetaTable = `CREATE TABLE if not exists mysql.stats_meta (
+		version bigint(64) unsigned NOT NULL,
+		table_id bigint(64) NOT NULL,
+		distinct_ratio double unsigned NOT NULL DEFAULT 0,
+		count bigint(64) unsigned NOT NULL DEFAULT 0,
```
Comment (coocood) [inline bootstrap.go]: I think we need to add a `modify_count` column.

**Comment by coocood** [comment]
@hanfei1991 

`modify_count` can be used to detect potential stats inaccuracy.

File: executor/analyze.go [hunk 12]
```diff
@@ -16,6 +16,7 @@ package executor
 import (
 	"math/rand"
 
+	"fmt"
```
Comment (coocood) [inline executor/analyze.go]: use `goimports` to format import.

File: plan/statscache/statscache.go [hunk 13]
```diff
@@ -14,89 +14,109 @@
...
-	txn := ctx.Txn()
-	if txn == nil {
-		err := ctx.ActivePendingTxn()
-		if err != nil || ctx.Txn() == nil {
-			return statistics.PseudoTable(tblInfo)
+	for _, row := range rows {
+		version, tableID := row.Data[0].GetUint64(), row.Data[1].GetInt64()
+		table, ok := is.TableByID(tableID)
+		if !ok {
+			log.Warnf("Unknown table ID %d in stats meta table, maybe it has been dropped", tableID)
```
Comment (coocood) [inline plan/statscache/statscache.go]: It's very common that a table has been dropped, should not log warning here.
This is an error only when we inserted a wrong table ID, so I think debug level log is enough.

File: plan/statscache/statscache_test.go [hunk 14]
```diff
@@ -33,44 +35,39 @@ var _ = Suite(&testStatsCacheSuite{})
...
+	testKit.MustExec("analyze table t")
+	statsTbl = statscache.GetStatisticsTableCache(tableInfo)
+	c.Assert(statsTbl.Pseudo, IsFalse)
+	testKit.MustExec("create index idx_t on t(c1)")
+	is = do.InfoSchema()
+	tbl, err = is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	tableInfo = tbl.Meta()
+	statsTbl = statscache.GetStatisticsTableCache(tableInfo)
+	c.Assert(statsTbl.Pseudo, IsTrue)
```
Comment (coocood) [inline plan/statscache/statscache_test.go]: When the schema has changed, we should not get a Pseudo table stats.
If the stats doesn't match the new schema, we should adapt the stats for the new schema.
Or we will keep using the pseudo stats for a long time.

Comment (hanfei1991) [inline plan/statscache/statscache_test.go]: We should adapt the stats after doing ddl. I will implement it in future.

File: bootstrap.go [hunk 15]
```diff
@@ -120,7 +120,7 @@ const (
 	CreateStatsMetaTable = `CREATE TABLE if not exists mysql.stats_meta (
 		version bigint(64) unsigned NOT NULL,
 		table_id bigint(64) NOT NULL,
-		distinct_ratio double unsigned NOT NULL DEFAULT 0,
+		modify_count bigint(64) NOT NULL DEFAULT 0,
```
Comment (coocood) [inline bootstrap.go]: Why removed `distinct_ratio`?

Comment (hanfei1991) [inline bootstrap.go]: distinct_ratio is for every column, which should be added to another table.

**Comment by coocood** [comment]
LGTM

**Pull request merged**